File: src/game/generator.ts
import { DAILY_ROUNDS, DAILY_SEED_VERSION } from "../config";
import { dailySeed, RNG } from "./rng";
import { difficultyT, lerp, roundDurationMs } from "./scoring";
import type { DiffType, Family, Round, ShapeParams } from "./types";

const COLOR_DIFFS: ReadonlySet<DiffType> = new Set(["hue", "saturation"]);

const FAMILIES: readonly Family[] = [
  "card",
  "circle",
  "triangle",
  "diamond",
  "hexagon",
  "star",
  "plus",
  "ring",
  "capsule",
  "chevron",
  "arc",
  "bars",
  "dots",
  "slash",
  "frame",
  "pentagon",
];

const FAMILY_DIFFS: Record<Family, readonly DiffType[]> = {
  card: ["hue", "saturation", "scale", "rotation", "stroke", "micro-detail", "offset"],
  circle: ["hue", "saturation", "scale", "stroke", "micro-detail", "offset", "count"],
  triangle: ["hue", "saturation", "scale", "rotation", "stroke", "micro-detail", "offset", "mirror"],
  diamond: ["hue", "saturation", "scale", "rotation", "stroke", "micro-detail", "offset"],
  hexagon: ["hue", "saturation", "scale", "rotation", "stroke", "micro-detail", "offset"],
  star: ["hue", "saturation", "scale", "rotation", "stroke", "micro-detail", "offset"],
  plus: ["hue", "saturation", "scale", "rotation", "stroke", "micro-detail", "offset"],
  ring: ["hue", "saturation", "scale", "stroke", "micro-detail", "offset"],
  capsule: ["hue", "saturation", "scale", "rotation", "stroke", "micro-detail", "offset"],
  chevron: ["hue", "saturation", "scale", "rotation", "stroke", "offset", "mirror"],
  arc: ["hue", "saturation", "scale", "rotation", "stroke", "offset", "mirror"],
  bars: ["hue", "saturation", "scale", "rotation", "stroke", "offset", "count"],
  dots: ["hue", "saturation", "scale", "rotation", "offset", "count"],
  slash: ["hue", "saturation", "scale", "rotation", "stroke", "offset", "mirror"],
  frame: ["hue", "saturation", "scale", "rotation", "stroke", "micro-detail", "offset"],
  pentagon: ["hue", "saturation", "scale", "rotation", "stroke", "micro-detail", "offset"],
};

function cloneParams(p: ShapeParams): ShapeParams {
  return { ...p };
}

function pickDiff(
  rng: RNG,
  family: Family,
  roundIndex: number,
  recent: DiffType[],
): DiffType {
  let pool = [...FAMILY_DIFFS[family]];
  const lastTwoColor =
    recent.length >= 2 && recent.slice(-2).every((d) => COLOR_DIFFS.has(d));
  if (lastTwoColor) {
    pool = pool.filter((d) => !COLOR_DIFFS.has(d));
  }
  if (roundIndex < 5) {
    const structural = pool.filter((d) => !COLOR_DIFFS.has(d));
    if (structural.length) pool = structural;
  }
  if (roundIndex < 4) {
    const obvious = pool.filter((d) =>
      d === "rotation" || d === "scale" || d === "count" || d === "offset" || d === "mirror",
    );
    if (obvious.length) pool = obvious;
  }
  if (pool.length === 0) {
    pool = [...FAMILY_DIFFS[family]];
  }
  return rng.pick(pool);
}

function baseParams(rng: RNG, family: Family, t: number): ShapeParams {
  const chromatic = rng.chance(0.42);
  const fillH = rng.range(0, 360);
  const fillS = chromatic ? rng.range(38, 68) : rng.range(6, 16);
  const fillL = chromatic ? rng.range(58, 74) : rng.range(88, 94);
  const countBase =
    family === "dots" ? 4 + rng.int(2) : family === "bars" ? 3 + rng.int(2) : family === "circle" ? 3 + rng.int(2) : 3;
  return {
    family,
    fillH,
    fillS,
    fillL,
    strokeOn: family === "frame" || family === "ring" || rng.chance(0.28),
    strokeW: lerp(5.5, 2.2, t),
    rotation: family === "circle" || family === "ring" || t < 0.22 ? 0 : rng.range(-10, 10),
    scale: 1,
    offsetX: 0,
    offsetY: 0,
    mirror: false,
    count: countBase,
    micro: false,
    microSize: lerp(9, 3.2, t),
  };
}

function applyDiff(
  rng: RNG,
  base: ShapeParams,
  diff: DiffType,
  t: number,
): ShapeParams {
  const odd = cloneParams(base);
  const sign = rng.sign();
  switch (diff) {
    case "hue":
      odd.fillH = (odd.fillH + sign * lerp(46, 7, t) + 360) % 360;
      break;
    case "saturation":
      odd.fillS = clamp(odd.fillS + sign * lerp(34, 8, t), 4, 90);
      break;
    case "scale":
      odd.scale = lerp(1.42, 1.055, t);
      break;
    case "rotation":
      odd.rotation += sign * lerp(34, 3.4, t);
      break;
    case "stroke":
      if (base.strokeOn) {
        odd.strokeW = Math.max(0.8, base.strokeW + sign * lerp(4.2, 1.15, t));
      } else {
        odd.strokeOn = true;
        odd.strokeW = lerp(6.5, 1.7, t);
      }
      break;
    case "micro-detail":
      odd.micro = true;
      odd.microSize = lerp(10, 3.1, t);
      break;
    case "offset": {
      const mag = lerp(20, 2.8, t);
      if (rng.chance(0.5)) odd.offsetX = sign * mag;
      else odd.offsetY = sign * mag;
      break;
    }
    case "count":
      odd.count = Math.max(1, base.count + (rng.chance(0.5) ? 1 : -1));
      if (odd.count === base.count) odd.count = base.count + 1;
      break;
    case "mirror":
      odd.mirror = !base.mirror;
      break;
    default: {
      const _exhaustive: never = diff;
      return _exhaustive;
    }
  }
  return odd;
}

function clamp(n: number, lo: number, hi: number): number {
  return Math.min(hi, Math.max(lo, n));
}

export function generateRound(
  rng: RNG,
  roundIndex: number,
  recent: DiffType[],
): Round {
  const t = difficultyT(roundIndex);
  let family = rng.pick(FAMILIES);
  if (roundIndex < 4) {
    family = rng.pick(["chevron", "slash", "bars", "triangle", "star", "card", "arc"]);
  }
  const diffType = pickDiff(rng, family, roundIndex, recent);
  const oddIndex = rng.int(3) as 0 | 1 | 2;
  const base = baseParams(rng, family, t);
  const odd = applyDiff(rng, base, diffType, t);
  const items: [ShapeParams, ShapeParams, ShapeParams] = [
    cloneParams(base),
    cloneParams(base),
    cloneParams(base),
  ];
  items[oddIndex] = odd;
  return {
    family,
    diffType,
    oddIndex,
    items,
    durationMs: roundDurationMs(roundIndex),
  };
}

export function generateRun(seed: number, count: number): Round[] {
  const rng = new RNG(seed);
  const recent: DiffType[] = [];
  const rounds: Round[] = [];
  for (let i = 0; i < count; i++) {
    const round = generateRound(rng, i, recent);
    recent.push(round.diffType);
    if (recent.length > 6) recent.shift();
    rounds.push(round);
  }
  return rounds;
}

export function generateDaily(dateId: string): Round[] {
  return generateRun(dailySeed(dateId, DAILY_SEED_VERSION), DAILY_ROUNDS);
}

export function isColorDiff(d: DiffType): boolean {
  return COLOR_DIFFS.has(d);
}

File: android/app/src/main/java/app/skew/android/game/Generator.kt
package app.skew.android.game

import app.skew.android.Config
import kotlin.math.max

private val COLOR_DIFFS = setOf(DiffType.HUE, DiffType.SATURATION)

private val FAMILIES = listOf(
    Family.CARD, Family.CIRCLE, Family.TRIANGLE, Family.DIAMOND, Family.HEXAGON,
    Family.STAR, Family.PLUS, Family.RING, Family.CAPSULE, Family.CHEVRON,
    Family.ARC, Family.BARS, Family.DOTS, Family.SLASH, Family.FRAME, Family.PENTAGON,
)

private val EARLY_FAMILIES = listOf(
    Family.CHEVRON, Family.SLASH, Family.BARS, Family.TRIANGLE, Family.STAR, Family.CARD, Family.ARC,
)

private val FAMILY_DIFFS: Map<Family, List<DiffType>> = mapOf(
    Family.CARD to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.ROTATION, DiffType.STROKE, DiffType.MICRO_DETAIL, DiffType.OFFSET),
    Family.CIRCLE to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.STROKE, DiffType.MICRO_DETAIL, DiffType.OFFSET, DiffType.COUNT),
    Family.TRIANGLE to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.ROTATION, DiffType.STROKE, DiffType.MICRO_DETAIL, DiffType.OFFSET, DiffType.MIRROR),
    Family.DIAMOND to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.ROTATION, DiffType.STROKE, DiffType.MICRO_DETAIL, DiffType.OFFSET),
    Family.HEXAGON to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.ROTATION, DiffType.STROKE, DiffType.MICRO_DETAIL, DiffType.OFFSET),
    Family.STAR to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.ROTATION, DiffType.STROKE, DiffType.MICRO_DETAIL, DiffType.OFFSET),
    Family.PLUS to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.ROTATION, DiffType.STROKE, DiffType.MICRO_DETAIL, DiffType.OFFSET),
    Family.RING to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.STROKE, DiffType.MICRO_DETAIL, DiffType.OFFSET),
    Family.CAPSULE to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.ROTATION, DiffType.STROKE, DiffType.MICRO_DETAIL, DiffType.OFFSET),
    Family.CHEVRON to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.ROTATION, DiffType.STROKE, DiffType.OFFSET, DiffType.MIRROR),
    Family.ARC to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.ROTATION, DiffType.STROKE, DiffType.OFFSET, DiffType.MIRROR),
    Family.BARS to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.ROTATION, DiffType.STROKE, DiffType.OFFSET, DiffType.COUNT),
    Family.DOTS to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.ROTATION, DiffType.OFFSET, DiffType.COUNT),
    Family.SLASH to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.ROTATION, DiffType.STROKE, DiffType.OFFSET, DiffType.MIRROR),
    Family.FRAME to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.ROTATION, DiffType.STROKE, DiffType.MICRO_DETAIL, DiffType.OFFSET),
    Family.PENTAGON to listOf(DiffType.HUE, DiffType.SATURATION, DiffType.SCALE, DiffType.ROTATION, DiffType.STROKE, DiffType.MICRO_DETAIL, DiffType.OFFSET),
)

private fun pickDiff(rng: Rng, family: Family, roundIndex: Int, recent: List<DiffType>): DiffType {
    var pool = FAMILY_DIFFS.getValue(family).toMutableList()
    val lastTwoColor = recent.size >= 2 && recent.takeLast(2).all { it in COLOR_DIFFS }
    if (lastTwoColor) pool = pool.filter { it !in COLOR_DIFFS }.toMutableList()
    if (roundIndex < 5) {
        val structural = pool.filter { it !in COLOR_DIFFS }
        if (structural.isNotEmpty()) pool = structural.toMutableList()
    }
    if (roundIndex < 4) {
        val obvious = pool.filter {
            it == DiffType.ROTATION || it == DiffType.SCALE || it == DiffType.COUNT ||
                it == DiffType.OFFSET || it == DiffType.MIRROR
        }
        if (obvious.isNotEmpty()) pool = obvious.toMutableList()
    }
    if (pool.isEmpty()) pool = FAMILY_DIFFS.getValue(family).toMutableList()
    return rng.pick(pool)
}

private fun baseParams(rng: Rng, family: Family, t: Double): ShapeParams {
    val chromatic = rng.chance(0.42)
    val fillH = rng.range(0.0, 360.0)
    val fillS = if (chromatic) rng.range(38.0, 68.0) else rng.range(6.0, 16.0)
    val fillL = if (chromatic) rng.range(58.0, 74.0) else rng.range(88.0, 94.0)
    val countBase = when (family) {
        Family.DOTS -> 4 + rng.int(2)
        Family.BARS -> 3 + rng.int(2)
        Family.CIRCLE -> 3 + rng.int(2)
        else -> 3
    }
    return ShapeParams(
        family = family,
        fillH = fillH,
        fillS = fillS,
        fillL = fillL,
        strokeOn = family == Family.FRAME || family == Family.RING || rng.chance(0.28),
        strokeW = lerp(5.5, 2.2, t),
        rotation = if (family == Family.CIRCLE || family == Family.RING || t < 0.22) 0.0 else rng.range(-10.0, 10.0),
        scale = 1.0,
        offsetX = 0.0,
        offsetY = 0.0,
        mirror = false,
        count = countBase,
        micro = false,
        microSize = lerp(9.0, 3.2, t),
    )
}

private fun clamp(n: Double, lo: Double, hi: Double): Double = minOf(hi, maxOf(lo, n))

private fun applyDiff(rng: Rng, base: ShapeParams, diff: DiffType, t: Double): ShapeParams {
    var odd = base.copy()
    val sign = rng.sign()
    when (diff) {
        DiffType.HUE -> odd = odd.copy(fillH = (odd.fillH + sign * lerp(46.0, 7.0, t) + 360) % 360)
        DiffType.SATURATION -> odd = odd.copy(fillS = clamp(odd.fillS + sign * lerp(34.0, 8.0, t), 4.0, 90.0))
        DiffType.SCALE -> odd = odd.copy(scale = lerp(1.42, 1.055, t))
        DiffType.ROTATION -> odd = odd.copy(rotation = odd.rotation + sign * lerp(34.0, 3.4, t))
        DiffType.STROKE -> {
            odd = if (base.strokeOn) {
                odd.copy(strokeW = max(0.8, base.strokeW + sign * lerp(4.2, 1.15, t)))
            } else {
                odd.copy(strokeOn = true, strokeW = lerp(6.5, 1.7, t))
            }
        }
        DiffType.MICRO_DETAIL -> odd = odd.copy(micro = true, microSize = lerp(10.0, 3.1, t))
        DiffType.OFFSET -> {
            val mag = lerp(20.0, 2.8, t)
            odd = if (rng.chance(0.5)) odd.copy(offsetX = sign * mag) else odd.copy(offsetY = sign * mag)
        }
        DiffType.COUNT -> {
            var c = max(1, base.count + if (rng.chance(0.5)) 1 else -1)
            if (c == base.count) c = base.count + 1
            odd = odd.copy(count = c)
        }
        DiffType.MIRROR -> odd = odd.copy(mirror = !base.mirror)
    }
    return odd
}

fun generateRound(rng: Rng, roundIndex: Int, recent: List<DiffType>): Round {
    val t = difficultyT(roundIndex)
    var family = rng.pick(FAMILIES)
    if (roundIndex < 4) family = rng.pick(EARLY_FAMILIES)
    val diffType = pickDiff(rng, family, roundIndex, recent)
    val oddIndex = rng.int(3)
    val base = baseParams(rng, family, t)
    val odd = applyDiff(rng, base, diffType, t)
    val items = mutableListOf(base.copy(), base.copy(), base.copy())
    items[oddIndex] = odd
    return Round(
        family = family,
        diffType = diffType,
        oddIndex = oddIndex,
        items = items,
        durationMs = roundDurationMs(roundIndex),
    )
}

fun generateRun(seed: UInt, count: Int): List<Round> {
    val rng = Rng(seed)
    val recent = mutableListOf<DiffType>()
    val rounds = mutableListOf<Round>()
    for (i in 0 until count) {
        val round = generateRound(rng, i, recent)
        recent.add(round.diffType)
        if (recent.size > 6) recent.removeAt(0)
        rounds.add(round)
    }
    return rounds
}

fun generateDaily(dateId: String): List<Round> =
    generateRun(dailySeed(dateId, Config.DAILY_SEED_VERSION), Config.DAILY_ROUNDS)

fun isColorDiff(d: DiffType): Boolean = d in COLOR_DIFFS
